tools/s3bulk/main.go

```diff
@@ -0,0 +1,237 @@
+	_ = filepath.Walk(searchDir, func(path string, f os.FileInfo, err error) error {
+
+		// check if it's a file/directory (we just want files)
+		file, err := os.Open(path)
+		if err != nil {
+			return nil
+		}
+
+		defer file.Close() // close file handle on return
+
+		fi, err := file.Stat()
+		if fi.IsDir() {
+			return nil
+		}
+
+		path = strings.Replace(path, searchDir, "", 1)
+		file_channel <- path // add file to the work channel (queue)
+		return nil
+	})
+
+	// add num_workers empty files on as termination signal to them
+	for i := 0; i < num_workers; i++ {
+		file_channel <- ""
+	}
+}
+
+// upload function for workers
+// uploads a given file to s3
+func (worker *Worker) upload(file string) (string, error) {
+	res, err := exec.Command("hab", "pkg", "info", "-j", fmt.Sprintf("%s%s", worker.SourceDir, file)).Output()
+
+	if err != nil {
+		panic(err)
+	}
+
+	var pkgInfo PkgInfo
+
+	err = json.Unmarshal(res, &pkgInfo)
+	if err != nil {
+		panic(err)
+	}
+
+	// s3 destination file path
+	var platform string
+
+	if strings.Contains(file, "windows") {
+		platform = "windows"
+	} else {
+		platform = "linux"
+	}
+
+	filename := strings.Split(file, "/")
+	destfile := fmt.Sprintf("/%s/%s/%s/%s/x86_64/%s/%s", pkgInfo.Origin, pkgInfo.Name, pkgInfo.Version, pkgInfo.Release, platform, filename[len(filename)-1])
+	worker.println("uploading to " + destfile)
+
+	// open and read file
+	f, err := os.Open(worker.SourceDir + file)
+
+	if err != nil {
+		return "Couldn't open file", err
+	}
+
+	defer f.Close()
+	fileInfo, _ := f.Stat()
+	var size = fileInfo.Size()
+	buffer := make([]byte, size)
+	f.Read(buffer)
+	fileBytes := bytes.NewReader(buffer)
+
+	params := &s3.PutObjectInput{
+		Bucket: aws.String(worker.Bucket),
+		Key:    aws.String(destfile),
+		Body:   fileBytes,
+		ACL:    aws.String(worker.Acl),
+	}
+
+
+	// try the actual s3 upload
+	resp, err := worker.Svc.PutObject(params)
+	if err != nil {
+		return "", err
+	} else {
+		return awsutil.StringValue(resp), nil
+	}
+}
+
+// doUploads function for workers
+//
+// reads from the file channel (queue),
+// calls upload function for each,
+// then moves uploaded files to worker.DestDir
+func (worker *Worker) doUploads() {
+
+	defer worker.Wg.Done() // notify parent when I complete
+
+	worker.println("doUploads() started")
+
+	// loop until I receive "" as a termination signal
+	for {
+		file := <-worker.File_channel
+		if file == "" {
+			break
+		}
+		worker.println("File to upload: " + file)
+		response, err := worker.upload(file)
+		if err != nil {
+			worker.println("error uploading" + file + ": " + response + " " + err.Error())
+		} else {
+			worker.println(response)
+			// make destination directory if needed
+			filename := path.Base(file)
+			directory := strings.Replace(file, "/"+filename, "", 1)
+			os.MkdirAll(worker.DestDir+directory, 0775)
+			// move file
+			os.Rename(worker.SourceDir+file, worker.DestDir+file)
```

Comment:
I think it would be  better to just leave the artifacts in place as is instead of a move right away.. do cleanup separately..eg, fir on-prem allow operator to cleanup after everything is verified to be working well.